### PR TITLE
[F] Range Input Component

### PR DESCRIPTION
**Pull request by drapisarda**
A first attempt! Let me know what you think of it!

The main idea is to have some HTML elements under a real and transparent input[type="range"] and then maintain the sync between UI and input value. 

It also works using the keyboard!

### CODE REVIEW

File: src/components/form/HdRange.vue [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    },
+    maxValue() {
+      this.updateRangeDecoration();
+    },
+  },
+  mounted() {
+    this.input = this.$el.querySelector('input');
+    this.progress = this.$el.querySelector('.range__progress');
+    this.progress.style.width = '1px';
+    this.thumb = this.$el.querySelector('.range__thumb');
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: I think it's better to use [$refs](https://vuejs.org/v2/api/#ref) instead of `this.$el.querySelector`

File: src/components/form/HdRange.vue [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+      &::-ms-track,
+      &::-ms-fill-lower,
+      &::-ms-fill-upper {
+        @include track-style;
+      }
+    }
+
+    &__decoration {
+      $decoration : &;
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: Doesn't seem like you're using this variable somewhere

File: src/stories/HdRange.stories.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      rangeStep: {
+        type: Number,
+        default: number('Step', 1),
+      },
+    },
+    methods: {
+      getValue() {
+        console.log('lol');
+      },
+    },
```
Comment (ilyasmez) [inline src/stories/HdRange.stories.js]: This is not used, maybe you can change it to print the value:
```js
watch: {
   value(value) {
     console.log('value', value); // or using the action addon
  },
},
```

File: src/components/form/HdRange.vue [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+    updateRangeDecoration() {
+      if (this.input.value > this.maxValue) {
+        this.input.value = this.maxValue;
+      }
+
+      if (this.input.value < this.minValue) {
+        this.input.value = this.minValue;
+      }
+
+      this.displayValue = this.input.value;
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: `displayValue` is not defined or used, this line can be removed.
Instead it seems like the right place to emit the `input` event:
```
this.$emit('input', parseFloat(this.input.value));
```

File: src/components/form/HdRange.vue [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    }
+
+    this.updateRangeDecoration();
+  },
+  computed: {
+  },
+  methods: {
+    updateRangeDecoration() {
+      if (this.input.value > this.maxValue) {
+        this.input.value = this.maxValue;
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: (Same this for the if statement bellow)

Instead of assigning to `this.input.value`, just emit an event `input` with the value you need, like so:
```js
this.$emit('input', this.maxValue);
```

v-model of the component (the one set on the parent) will take care of the binding :)

File: src/stories/HdRange.stories.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  .addDecorator(FormWrapper)
+  .add('default 🎛', () => ({
+    components: { HdRange },
+    template: `
+      <div>
+        <HdRange
+          :minValue=minValue
+          :maxValue=maxValue
+          :rangeStep=rangeStep
+          :value="value"
```
Comment (ilyasmez) [inline src/stories/HdRange.stories.js]: To keep the same api for the form components (necessary for HdDynamicForm), we prefer to use `v-model`.
```suggestion
          v-model="value"
```

File: src/components/form/HdRange.vue [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+      default: 0,
+    },
+    rangeStep: {
+      type: Number,
+      default: 1,
+    },
+  },
+  watch: {
+    value() {
+      this.input.value = this.value;
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: This is not needed it when using `v-model`.

File: src/components/form/HdRange.vue (hunk 1 above)
Comment (viniciuskneves) [reply src/components/form/HdRange.vue]: Agree

File: src/components/form/HdRange.vue [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+      this.updateRangeDecoration();
```
Comment (jpDenisO) [inline src/components/form/HdRange.vue]: the method `updateRangeDecoration` is already set as callback for @input and @change, wouldn't it get triggered x3 on each change?

Comment (jpDenisO) [reply src/components/form/HdRange.vue]: I personally would keep it only here at the watcher, and leave the events for callbacks passed from a parent component

File: src/components/form/HdRange.vue (hunk 1 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: From the docs I thought it should be used just for components, but I see it's not. I go for it

File: src/components/form/HdRange.vue (hunk 4 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: Ok. Do we need to emit this event? Doesn't it emit by itself?

Comment (drapisarda) [reply src/components/form/HdRange.vue]: Also, I made a method called _normalizeValue_ . It's always used before the uodateRangeDecoration but it's more descriptive.

File: src/components/form/HdRange.vue (hunk 5 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: Got it!

File: src/components/form/HdRange.vue (hunk 4 above)
Comment (ilyasmez) [reply src/components/form/HdRange.vue]: Yes, we need to emit the event, so the parent can know about the changes.

File: src/components/form/HdRange.vue (hunk 8 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: I had to set it on both change and input because IE11 emits change instead of input during the drag. So I have to keep them both. Unfortunately, I can not make it work at the watcher!

File: src/stories/HdRange.stories.js (hunk 6 above)
Comment (drapisarda) [reply src/stories/HdRange.stories.js]: I did not found a way to do it. I have to validate that value, so I can not use v-modal in the story, otherwise I get this 
image

File: src/components/form/HdRange.vue [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+      const thumbWidthOffset = this.thumb.offsetWidth * valuePercentage / 100;
+      this.thumb.style.transform = `translateX(${valuePercentageInputWidthPixels - thumbWidthOffset}px)`;
+    },
+    handleFocus() {
+      this.isActive = true;
+    },
+    handleBlur() {
+      this.isActive = false;
+    },
+    currentValue() {
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: You're using this method as a modal (`v-modal`)? I'm not sure that will work. A modal needs to be settable and gettable.
I suggest adding `currentValue` as a data property, renaming this method and calling it in the beginning of `updateRangeDecoration()`

File: src/components/form/HdRange.vue [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+      :name="name"
+      ref="input"
+      :v-model="currentValue"
+      :required="required"
+      :disabled="disabled"
+      :min="minValue"
+      :max="maxValue"
+      :step="rangeStep"
+      @input="updateRangeDecoration"
+      @change="isLegacy ? updateRangeDecoration : null"
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: `v-modal` handles this (with the edges cases that you mentioned in the other comment as well).
If you make the changes as I suggested in the other comment, watching `currentValue` will be enough.

File: src/stories/HdRange.stories.js (hunk 6 above)
Comment (drapisarda) [reply src/stories/HdRange.stories.js]: I'm using the _v-model_ on the input element. The component still have _value_ as a prop.

File: src/components/form/HdRange.vue (hunk 9 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: I was using it to validate the value, but the range element input value does it by itself. So I remvove this.

File: src/components/form/HdRange.vue (hunk 10 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: Let's sync on this.

File: src/components/form/HdRange.vue [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+      v-model="currentValue"
+      :id="name"
+      :name="name"
+      :required="required"
+      :disabled="disabled"
+      :min="minValue"
+      :max="maxValue"
+      :step="rangeStep"
+      @input="updateRangeDecoration"
+      @change="isLegacy ? updateRangeDecoration : null"
```
Comment (viniciuskneves) [inline src/components/form/HdRange.vue]: I would add a test for this guy as well. As if `isLegacy` is set it won't call `updateRangeDecoration`.

Comment (viniciuskneves) [reply src/components/form/HdRange.vue]: You could also add it to the story so we could play with it as well.

Comment (drapisarda) [reply src/components/form/HdRange.vue]: Nice! I wrote the test and got that that feature was not working. I made it differently and works great. That's TDD's best flavour. Thank you, man.

I also wrote tests for the other listeners. Big Up!

**Comment by viniciuskneves** [comment]
Hey @drapisarda as I mentioned in Slack, I'm going to review it after PEW, unless you really need this component for your current development then let me know!

File: src/components/form/HdRange.vue [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+
+      this.currentValue = this.currentValue - (this.currentValue % this.rangeStep);
+
+      this.$emit('input', parseFloat(this.currentValue));
+    },
+    updateRangeDecoration() {
+      this.normalizeAndEmit();
+
+      // the percentage of the value, between max and min. I.e. : 15 is the 50% between 10 and 20
+      const valuePercentage = this.currentValue / ((this.$props.maxValue - this.$props.minValue) / 100);
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: The current implementation will work properly **only** if the minValue is `0`, this should fix it:
```suggestion
      const valuePercentage = (this.currentValue - this.minValue) / (this.maxValue - this.minValue) * 100;
```

And yeah, you don't have to use `$props`, (not like in some other frameworks, react... ehm) :D

File: src/components/form/HdRange.vue [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+      v-model="currentValue"
+      :id="name"
+      :name="name"
+      :required="required"
+      :disabled="disabled"
+      :min="minValue"
+      :max="maxValue"
+      :step="rangeStep"
+      @input="updateRangeDecoration"
+      @change="handleChange"
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: Now that we are using `v-model`, we don't have to handle these events manually. We can remove them.

File: src/components/form/HdRange.vue [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+    handleBlur() {
+      this.isActive = false;
+    },
+    handleChange() {
+      if (!this.isLegacy) {
+        return;
+      }
+
+      this.updateRangeDecoration();
+    },
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: We won't need this one as well.

File: src/components/form/HdRange.vue [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+      default: 100,
+    },
+    rangeStep: {
+      type: Number,
+      default: 1,
+    },
+    isLegacy: {
+      type: Boolean,
+      default: false,
+    },
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: This one can be removed as well.

File: src/components/form/HdRange.vue [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+    maxValue() {
+      this.updateRangeDecoration();
+    },
+    rangeStep() {
+      this.updateRangeDecoration();
+    },
+    value() {
+      this.currentValue = this.value;
+      this.updateRangeDecoration();
+    },
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: Here we can make use of the modal (that's how we won't need to handle the change/input events ourselves):

```js
value() {
  this.currentValue = this.value;
},
currentValue() {
  this.updateRangeDecoration();
},
```

File: src/components/form/HdRange.vue [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+    normalizeAndEmit() {
+      if (this.currentValue > this.maxValue) {
+        this.currentValue = this.maxValue;
+      } else if (this.currentValue < this.minValue) {
+        this.currentValue = this.minValue;
+      }
+
+      this.currentValue = this.currentValue - (this.currentValue % this.rangeStep);
+
+      this.$emit('input', parseFloat(this.currentValue));
```
Comment (ilyasmez) [inline src/components/form/HdRange.vue]: Small optional thing:
You don't have to parse the value to float yourself you can use [`v-modal.number`](https://vuejs.org/v2/guide/forms.html#number) (it does exactly the same thing, I only find it cleaner, and it's always nice to parse as early as possible).

File: src/components/form/HdRange.vue [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+      this.updateRangeDecoration();
+    },
+    value() {
+      this.currentValue = this.value;
+      this.updateRangeDecoration();
+    },
+  },
+  mounted() {
+    this.progress = this.$refs.progress;
+    this.progress.style.width = '1px';
```
Comment (SinisaG) [inline src/components/form/HdRange.vue]: Maybe a comment why this is needed?

File: src/components/form/HdRange.vue (hunk 12 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: Supper man

File: src/components/form/HdRange.vue (hunk 13 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: It does not work without those!

File: src/components/form/HdRange.vue (hunk 15 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: I need it to handle the change event on IE11

File: src/components/form/HdRange.vue (hunk 16 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: 🚀

File: tests/unit/components/form/HdRange.spec.js [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+
+  test('renders component', () => {
+    expect(wrapper.html()).toMatchSnapshot();
+  });
+
+  test('renders with preselected value', () => {
+    expect(parseFloat(wrapper.find('input').element.value)).toBe(initialValue);
+  });
+
+  test('manages the max value', () => {
```
Comment (SinisaG) [inline tests/unit/components/form/HdRange.spec.js]: Maybe a better wording here would be - `respects the limit of max value`

File: tests/unit/components/form/HdRange.spec.js [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+    wrapper.setProps({
+      value: maxValue * 2,
+    });
+
+    expect(parseFloat(wrapper.find('input').element.value)).toBe(maxValue);
+    expect(wrapper.emitted('input')[1][0]).toEqual(maxValue);
+  });
+
+  test('menages the min value', () => {
```
Comment (SinisaG) [inline tests/unit/components/form/HdRange.spec.js]: Similar here: `Limits the value to passed minimum value`

File: src/components/form/HdRange.vue (hunk 17 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: Very nice!

File: tests/unit/components/form/HdRange.spec.js [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+    wrapper.setProps({
+      isLegacy: false,
+    });
+
+    wrapper.find('input').trigger('change');
+
+    expect(mockedUpdateRangeDecoration).toHaveBeenCalledTimes(0);
+  });
+
+  test('input focus emit is managed', () => {
```
Comment (SinisaG) [inline tests/unit/components/form/HdRange.spec.js]: If I get it right, we're testing here that onFocus is called when `focus` event happens? I would suggest updated wording here also e.g. `focusHanlder is called on focus`

File: tests/unit/components/form/HdRange.spec.js [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+    const mockedHandleFocus = jest.fn();
+    wrapper.setMethods({ handleFocus: mockedHandleFocus });
+
+    wrapper.find('input').trigger('focus');
+
+    expect(mockedHandleFocus).toHaveBeenCalledTimes(1);
+  });
+
+  test('input blur emit is managed', () => {
```
Comment (SinisaG) [inline tests/unit/components/form/HdRange.spec.js]: Respectively here: `blurHanlder is called on blur `

File: src/components/form/HdRange.vue [hunk 23]
```diff
@@ -0,0 +1,335 @@
...
+    },
+    value() {
+      this.currentValue = this.value;
+    },
+    currentValue() {
+      this.updateRangeDecoration();
+    },
+  },
+  mounted() {
+    // the progress bar needs a starting fixed width to be then transformed in updateRangeDecoration() with scaleX property
```
Comment (SinisaG) [inline src/components/form/HdRange.vue]: Thanks for comment. :+1:

File: src/components/form/HdRange.vue [hunk 24]
```diff
@@ -0,0 +1,329 @@
...
+    },
+    blurHandler() {
+      this.isActive = false;
+    },
+  },
+};
+</script>
+
+<style scoped lang="scss">
+.range {
```
Comment (SinisaG) [inline src/components/form/HdRange.vue]: I suggest to explicitly import `mixins.scss` here, so we slowly shift away from the patter of having them injected with webpack in parent project.

File: src/components/form/HdRange.vue [hunk 25]
```diff
@@ -0,0 +1,329 @@
...
+      :max="maxValue"
+      :step="rangeStep"
+      @focus="focusHandler"
+      @blur="blurHandler"
+    />
+    <div class="range__decoration" ref="decoration">
+      <div class="range__progress" ref="progress" />
+    </div>
+    <ul v-if="displayStepBullets" class="range__steps">
+      <li v-for="(steps, i) in stepsAmount" :key="i" />
```
Comment (SinisaG) [inline src/components/form/HdRange.vue]: This is the "problematic" part.

File: tests/unit/components/form/HdRange.spec.js [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+    wrapper = wrapperFactory();
+  });
+
+  test('renders component', () => {
+    expect(wrapper.html()).toMatchSnapshot();
+  });
+
+  test('renders with preselected value', () => {
+    expect(parseFloat(wrapper.find('input').element.value)).toBe(initialValue);
+  });
```
Comment (viniciuskneves) [inline tests/unit/components/form/HdRange.spec.js]: The second assertion could be inside the first test as you initialize the component with a value already.

File: tests/unit/components/form/HdRange.spec.js [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+
+    wrapper.find('input').trigger('focus');
+
+    expect(mockedFocusHandler).toHaveBeenCalledTimes(1);
+  });
+
+  test('When the focus event is emitted, the input is in an active status', () => {
+    wrapper.find('input').trigger('focus');
+
+    expect(wrapper.vm.isActive).toBe(true);
```
Comment (viniciuskneves) [inline tests/unit/components/form/HdRange.spec.js]: I think those two guys can also be inside the same test (if I'm missing some context, please let me know).

File: tests/unit/components/form/HdRange.spec.js [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+
+    wrapper.find('input').trigger('blur');
+
+    expect(mockedBlurHandler).toHaveBeenCalledTimes(1);
+  });
+
+  test('When the blue event is emitted, the input is in an non active status', () => {
+    wrapper.find('input').trigger('blur');
+
+    expect(wrapper.vm.isActive).toBe(false);
```
Comment (viniciuskneves) [inline tests/unit/components/form/HdRange.spec.js]: Same here

File: src/components/form/HdRange.vue [hunk 29]
```diff
@@ -0,0 +1,329 @@
...
+      default: 1,
+    },
+    value: {
+      type: Number,
+      default: 0,
+    },
+    displayStepBullets: {
+      type: Boolean,
+      default: true,
+    },
```
Comment (viniciuskneves) [inline src/components/form/HdRange.vue]: I would add a test to cover this property as it is not tested anywhere and apparently, as pointed out by @SinisaG, there is a bug for this guy.

File: src/components/form/HdRange.vue (hunk 25 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: I fixed it, giving the number of steps as integer division and give a relative padding-right to the unsorted list. The value of the padding-right is the proportion of module of the division on the range "size".

File: src/components/form/HdRange.vue (hunk 24 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: So, same for var?

File: tests/unit/components/form/HdRange.spec.js (hunk 26 above)
Comment (drapisarda) [reply tests/unit/components/form/HdRange.spec.js]: 👍

File: tests/unit/components/form/HdRange.spec.js (hunk 27 above)
Comment (drapisarda) [reply tests/unit/components/form/HdRange.spec.js]: I can not put them together. The _isActive_ value is set by the _focusHandler_ method. In the first test, I'm mocking that method, so the _isActive_ value is not affected. So, I can make just one test from this two, but I would, however, have to trigger _focus_ twice. So, it's pointless to have just one test.

File: tests/unit/components/form/HdRange.spec.js (hunk 28 above)
Comment (drapisarda) [reply tests/unit/components/form/HdRange.spec.js]: Same also here.

File: src/components/form/HdRange.vue (hunk 29 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: I do agree.

File: tests/unit/components/form/HdRange.spec.js (hunk 27 above)
Comment (viniciuskneves) [reply tests/unit/components/form/HdRange.spec.js]: How about not mocking it on the first test?
You can just spy the `focusHandler` so you can test it is being called and the `isActive` state is set as well.
What do you think?

Comment (drapisarda) [reply tests/unit/components/form/HdRange.spec.js]: I changed the order and it works good!

File: package.json [hunk 30]
```diff
@@ -87,5 +87,13 @@
       "vue-cli-service lint",
       "git add"
     ]
+  },
+  "prettier": {
```
Comment (viniciuskneves) [inline package.json]: I don't think we've prettier setup in this project. Do you intend to do so? I would definitely recommend to do in another PR as it might update a lot of files all at once.

File: src/components/form/HdRange.vue [hunk 31]
```diff
@@ -78,7 +78,12 @@ export default {
       };
     },
     stepsAmount() {
-      return 1 + (this.maxValue - this.minValue) / this.rangeStep;
+      const rangeSize = this.maxValue - this.minValue;
```
Comment (viniciuskneves) [inline src/components/form/HdRange.vue]: You can make `rangeSize` a computed property by itself.

File: src/stories/HdRange.stories.js [hunk 32]
```diff
@@ -58,4 +58,25 @@ storiesOf('Form/HdRange', module)
...
     },
+  }))
+  .add('without steps', () => ({
+    components: { HdRange },
+    template: `
+      <div>
+        <HdRange
+          :minValue="0"
+          :maxValue="100"
+          :rangeStep="10"
```
Comment (viniciuskneves) [inline src/stories/HdRange.stories.js]: It has steps and the story says it doesn't 🤔

Comment (drapisarda) [reply src/stories/HdRange.stories.js]: FUUUUU! Bad description. It was actually without bullets. I made both the story and make the display bullet optional to make it behave more like the native one.

File: src/components/form/HdRange.vue (hunk 31 above)
Comment (drapisarda) [reply src/components/form/HdRange.vue]: 👍

File: package.json (hunk 30 above)
Comment (drapisarda) [reply package.json]: 👍

**Comment by viniciuskneves** [comment]
Hey @drapisarda there are some conflicts in the `package-lock` file. You can run `npm install --package-lock-only` to solve it.
What I actually wanted to say is: don't force-push when there are reviews pending as it messes with the commit history and the comments itself. Just as a side note and suggestion of course.
Going to go through it once again today.
